src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.cs

```diff
@@ -1,97 +1,122 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
-using System.Linq;
+using System.Runtime.CompilerServices;
 using Microsoft.AspNetCore.Razor.Language.Syntax;
 
 namespace Microsoft.AspNetCore.Razor.Language.Legacy;
 
-internal static class LegacySyntaxNodeExtensions
+internal static partial class LegacySyntaxNodeExtensions
 {
-    private static readonly SyntaxKind[] TransitionSpanKinds = new SyntaxKind[]
+    private class SpanData
+    {
+        public SyntaxNode? Previous;
+        public bool PreviousComputed;
+        public SyntaxNode? Next;
+        public bool NextComputed;
+    }
+
+    /// <summary>
+    ///  Caches previous/next span result for a particular node. A conditional weak table
+    ///  is used to avoid adding fields to all syntax nodes.
+    /// </summary>
+    private static readonly ConditionalWeakTable<SyntaxNode, SpanData> s_spanDataTable = new();
+
+    private static readonly ISet<SyntaxKind> s_transitionSpanKinds = new HashSet<SyntaxKind>
```

Comment:
I would prefer to pull in the changes from https://github.com/dotnet/razor/pull/8234#issuecomment-1423095792 and make these immutable sets.

Reply:
I'd hoped to use immutable sets and was delighted when I saw that you were pulling the immutable collections in. I'll wait until rebase these changes once you're PR is merged.

Reply:
Update to immutable sets.